synapse/util/tracerutils.py

```diff
@@ -0,0 +1,309 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.d
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.import opentracing
+
+import logging
+import re
+from functools import wraps
+
+logger = logging.getLogger(__name__)
+
+
+def only_if_tracing(func):
+    """Executes the function only if we're tracing. Otherwise return.
+    Assumes the function wrapped may return None"""
+
+    @wraps(func)
+    def f(cls, *args, **kwargs):
+        if cls._opentracing:
+            return func(cls, *args, **kwargs)
+        else:
+            return
+
+    return f
+
+
+class TracerUtil(object):
```

Comment:
Actually, instead of making this global maybe we should just attach it to the HS object? So that you have:

```python
def __init__(self, hs):
    self._opentracing = hs.get_opentracing()

...

def foo(self):
    self._opentracing.start_span()
    ...
```

That way we don't have to mess around with classmethods and the like.

Reply:
Moving opentracing into the homeserver causes us a problem when trying to access opentracing within a request. Would it be ok to inject the opentracing reference across SynapseSite, SynapseRequestFactory and SyanpaseRequest?

Reply:
We discussed this in person and are going to thread it through those objects

Reply:
Oh boy, jeager_client silently starts a thread in order to report spans. We ran into the problem before that if we initialised jaeger during setup this new thread would be killed. We want to pass opentracing through the hs to various object during setup but we can only start a longliving jaeger client after setup. (During the `start` method in _base for example.)

Two solutions:
- Pass around an object which when use will resolve to our opentracing object later
- Go back to global state

I think going back to global state would be cleaner and easier to understand

Reply:
Ha, ok yes, you're right, let's go back to global state.

Reply:
Still need to change TracerUtil over to a class method